### ISSUE TITLE
channels should not hardcode usage of md5

### ISSUE THREAD

**Issue by wenottingham** [django/channels#575]
channels/sessions.py:
```
    # We hash the whole reply channel name and add a prefix, to fit inside 32B
    reply_name = reply_channel
    hashed = hashlib.md5(reply_name.encode("utf8")).hexdigest()
    session_key = "chn" + hashed[:29]
```

When the underlying OS is running in FIPS-140-2 mode (https://en.wikipedia.org/wiki/FIPS_140-2), MD5 is a disallowed cipher. This causes channels to crash.

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/channels/worker.py", line 120, in run
    consumer(message, **kwargs)
  File "/usr/lib/python2.7/site-packages/channels/sessions.py", line 54, in inner
    session = session_for_reply_channel(message.reply_channel.name)
  File "/usr/lib/python2.7/site-packages/channels/sessions.py", line 20, in session_for_reply_channel
    hashed = hashlib.md5(reply_name.encode("utf8")).hexdigest()
ValueError: error:060800A3:digital envelope routines:EVP_DigestInit_ex:disabled for fips
```

Ideally, this would use a newer hash algorithm that is allowed.

Tested on: CentOS 7

**Comment by andrewgodwin** [comment]
Huh, interesting. We're not using MD5 as a cryptographic cipher here but rather as an entropy extractor, and given that we trim the bytes off, it's definitely not secure anyway.

Can you provide details on which hashing functions are allowed? Can we switch to SHA-1, or do I need to look further up the chain?

**Comment by davidmarquis** [comment]
Looks like this document outlines the approved functions: http://csrc.nist.gov/publications/fips/fips140-2/fips1402annexa.pdf

Some hashing functions supported:  SHA-1, SHA-224, SHA-256, SHA-384, SHA-512, SHA-512/224
and SHA-512/256, SHA-3 variants (SHA3-224, SHA3-256, SHA3-384, SHA3-512)

**Comment by davidmarquis** [comment]
Also worth noting that SHA-1 has been "broken" by a Google team recently: https://shattered.it - but actual risk seems pretty low, especially for this use case (not cryptographic in nature)

**Comment by andrewgodwin** [comment]
Since I'm not too worried about the crypto properties involved here, I'll move to SHA-1 for now.

**Comment by andrewgodwin** [comment]
Fixed in 476d1500c6b0673863a995761c68c01b382139d2
